packages/channel-client/tests/fakes/fake-channel-provider.ts

```diff
@@ -174,7 +174,7 @@ export class FakeChannelProvider implements ChannelProviderInterface {
     // [assuming we're working with 2-participant channels for the time being]
     this.setState({
       ...latestState,
-      turnNum: bigNumberify(3).toString(),
+      turnNum: bigNumberify(2).toString(),
```

Comment:
☝️This is the subtle bug that causes issues.

Reply:
Hm, I've run into this in the past. What could we have done to make this more prevalent when this bug shows up?

Reply:
I think the console log error log is one thing we shouldn't ignore. The other thing we can do is add a test, now that we know.

Reply:
There is already a test, and it asserts the *current* behaviour. Also, the error in the console does not always appear.  Clearly a channel runs successfully some or most of the time.

Reply:
To be clear: I don't think there was a bug here. When you join the channel, you have state 0 from your opponent: you should sign state 1, safely fund the channel, wait for your opponent to sign state 2, and then sign state 3. The FakeProvider shortcuts some of this, but the joining party (participant `1`) should always be signing odd-numbered-turnNum states.

Reply:
Should calling `joinChannel` increase the turn number? Or the increase and signing happens after? Everything seems to work fine after I make this change.

Reply:
I think I found out what really happened.

The below are logs from my PR (without the turnNum = 2 fix) rebased on George's PR: https://github.com/statechannels/monorepo/pull/1164

Seeder log:

![image](https://user-images.githubusercontent.com/2715151/76368597-2bdc4c00-6307-11ea-9319-309448ffcad5.png)

Leecher log:

![image](https://user-images.githubusercontent.com/2715151/76368584-20892080-6307-11ea-8334-2c70a7704c06.png)

Seeder creates a channel, and sends a STOP event to leecher. Upon receiving STOP, leecher tries to transfer funds. At that moment, leecher hasn't even joined the channel yet. (For evidence, only after the 2nd STOP, we saw the `payment made...` message in the leecher log).

George's fix with a condition inside the `transferFund` isn't correct because it actually makes the leecher skip the first payment. That's why in the Seeder log, after the accepting the first payment,  `newSeederBalance` is still zero!

I think the safest fix is to hold the first STOP firing until leecher joins the channel. 

What do you think? @geoknee

Reply:
It's such a coincidence that, when I put in turnNum = 2 fix, first eager STOP event actually makes leecher to transfer fund successfully and increase the turnNum correctly to 3 before it joins the channel. That could mean some checks may be missing in fake channel provider.

Reply:
> I think the safest fix is to hold the first STOP firing until leecher joins the channel.

This would lead to the leecher being unchoked and able to download until they join the channel, which doesn't seem right.

Reply:
> This would lead to the leecher being unchoked and able to download until they join the channel, which doesn't seem right.

What I meant is seeder still chokes the wire but holds the first STOP until leecher joins the channel. STOP essentially equals `asking for payment` in our implementation. In other words, with this proposed fix, we decouple wire-choking and asking for payment. It does add little complexity to handle the initial state, but it is still reasonable? 🤷‍♂ 

The fix in https://github.com/statechannels/monorepo/pull/1164 results that, leecher joining the channel is equivalent to paying the first payment, therefore starts leeching right away. That doesn't seem right either.

Reply:
Yes I think this is reasonable. Let's implement that after merging this and #1164 . 

(We discussed that having a little bit of free leeching might not be so bad, but that it's generally nicer to not request a payment if you don't need one, rather than requesting the payment and have the leecher decide not to pay).

Reply:
Filed the new issue: https://github.com/statechannels/monorepo/issues/1178